app/Support/Modules/DatabaseActivator.php

```diff
@@ -0,0 +1,143 @@
+<?php
+
+namespace App\Support\Modules;
+
+use Illuminate\Config\Repository as Config;
+use Illuminate\Container\Container;
+use Nwidart\Modules\Contracts\ActivatorInterface;
+use Nwidart\Modules\Module;
+
+class DatabaseActivator implements ActivatorInterface
+{
+    /**
+     * Laravel config instance
+     *
+     * @var Config
+     */
+    private $config;
+
+    /**
+     * Module model instance
+     *
+     * @var Config
+     */
+    private $model;
+
+    /**
+     * Array of modules activation statuses
+     *
+     * @var array
+     */
+    private $modulesStatuses;
+
+    public function __construct(Container $app)
+    {
+        $this->config = $app['config'];
+        $this->modulesStatuses = $this->getModulesStatuses();
+        $this->model = (new \App\Models\Module());
+    }
+
+    /**
+     * Get modules statuses, from the database
+     *
+     * @return array
+     */
+    private function getModulesStatuses(): array
+    {
+        $modules = \App\Models\Module::all();
+        $retVal = [];
+        foreach ($modules as $i) {
+            $retVal[$i->name] = $i->enabled;
+        }
+        return $retVal;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function reset(): void
+    {
+        $this->model->truncate();
+        $this->modulesStatuses = [];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function enable(Module $module): void
+    {
+        $this->setActiveByName($module->getName(), true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function disable(Module $module): void
+    {
+        $this->setActiveByName($module->getName(), false);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasStatus(Module $module, bool $status): bool
+    {
+        if (!isset($this->modulesStatuses[$module->getName()])) {
+            return $status === false;
+        }
+
+        return $this->modulesStatuses[$module->getName()] === $status;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setActive(Module $module, bool $active): void
+    {
+        $this->setActiveByName($module->getName(), $active);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setActiveByName(string $name, bool $status): void
+    {
+        $this->modulesStatuses[$name] = $status;
+        $this->writeDB($name, $status);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function delete(Module $module): void
+    {
+        $module_name = $module->getName();
+        if (!isset($this->modulesStatuses[$module_name])) {
+            return;
+        }
+        unset($this->modulesStatuses[$module_name]);
+        $this->writeDB($module_name, false, 1);
+    }
+
+    /**
+     * Writes the activation statuses in a file, as json
+     *
+     * @param $name
+     * @param $status
+     * @param string $delete
+     */
+    private function writeDB($name, $status, $delete = ''): void
```

Comment:
Don't use a method like this, or the `$this->model`, just simplify this by calling the model methods directly (`\App\Model\Module::where(...)->delete()`, etc) from the caller. Having a proxy method like this is unwieldy and adds maintenance cost going forward, and then variable parameters, etc is not good.

Same thing in something like `hasStatus()`. We can optimize this later by adding a `ModuleRepository` and proxying the DB methods through that, and that will handle caching. For now, the methods being as 'live' as possible is best.

Reply:
Alright!

Reply:
Oh, that means I should connect `hasStatus()` method with DB and not statuses right?

Reply:
Yeah, for this whole file. Because later on, once this is working, we'll wire it as a repository and change the calls to use that. Then we get caching for free.

Reply:
I think the `hasStatus()` method is directly connected to DB since in `construct `method we gather data from the DB itself and it is using that data for getting status right?
Please correct me if I am wrong in understanding the caching part

Reply:
Don't worry about the caching part yet, just explaining the reasoning behind it. It just needs to use the model directly, not through the writeDB method here